File: .portkit/scripts/python/update-registry.py
import os
import re
import json
import argparse
from pathlib import Path

REGISTRY_PATH = Path(".portkit/addon-features-registry/feature-registry.json")

def scan_file(path, feature_map):
    try:
        with open(path, 'r', encoding='utf-8') as f:
            content = f.read()
    except:
        return

    # Find all feature-start tags
    matches = re.findall(r"//\s*feature-start:\s*([\w-]+)", content)
    # Also support python hash comments
    matches += re.findall(r"#\s*feature-start:\s*([\w-]+)", content)
    
    for feature in matches:
        if feature not in feature_map:
            feature_map[feature] = []
        if str(path) not in feature_map[feature]:
            feature_map[feature].append(str(path))

def get_changed_files():
    """Get list of modified and new files from git."""
    files = set()
    try:
        # 1. Modified tracked files (diff against HEAD)
        res = subprocess.run(['git', 'diff', '--name-only', 'HEAD'], capture_output=True, text=True)
        if res.returncode == 0:
            files.update(res.stdout.splitlines())

        # 2. Untracked (new) files
        res = subprocess.run(['git', 'ls-files', '--others', '--exclude-standard'], capture_output=True, text=True)
        if res.returncode == 0:
            files.update(res.stdout.splitlines())
    except Exception as e:
        print(f"Warning: Could not get git status for audit: {e}")
    
    # Filter for relevant extensions
    return [Path(f) for f in files if f.endswith(('.ts', '.tsx', '.js', '.py', '.md'))]

def main():
    parser = argparse.ArgumentParser(description="Scan codebase for feature tags.")
    parser.add_argument("--update", action="store_true", help="Update registry file")
    parser.add_argument("--audit", action="store_true", help="Check for changed files missing feature tags")
    parser.add_argument("--folder", help="Limit scan to specific folder")
    args = parser.parse_args()

    feature_map = {}
    
    import subprocess
    feature_map = {}
    
    try:
        # OPTIMIZATION: Use git grep to find ONLY files that contain our tags.
        # This avoids opening thousands of files that have no tags.
        # -I: Ignore binary
        # -l: Print filename only
        # "feature-start:" matches both // and # styles
        
        # Determine search path from args
        search_path = args.folder if args.folder else '.'
        
        cmd = ['git', 'grep', '-I', '-l', 'feature-start:', '--', search_path]
        
        result = subprocess.run(cmd, capture_output=True, text=True)
        
        if result.returncode == 0:
            files = result.stdout.splitlines()
            # Normalize paths to handle potential OS differences in git output (though usually forward slash)
            files = [f.strip() for f in files if f.strip()]
            
            for file_path in files:
                path = Path(file_path)
                if path.exists():
                    scan_file(path, feature_map)
        else:
            # git grep returns 1 if no matches found, which is fine
            if result.returncode != 1:
                print(f"git grep returned code {result.returncode}")

    except Exception as e:
        print(f"Error running git grep optimization: {e}. Falling back to slow walk.")
        # Fallback to os.walk if git isn't available
        for root, dirs, files in os.walk("."):
            if ".git" in dirs: dirs.remove(".git")
            if "node_modules" in dirs: dirs.remove("node_modules")
            if ".portkit-cache" in dirs: dirs.remove(".portkit-cache")
            
            for file in files:
                if file.endswith(('.ts', '.tsx', '.js', '.py', '.md')):
                    path = Path(root) / file
                    scan_file(path, feature_map)

    # Output or Update
    if args.update:
        if REGISTRY_PATH.exists():
            with open(REGISTRY_PATH, 'r') as f:
                registry = json.load(f)
        else:
            registry = {"features": {}, "metadata": {"version": "1.0"}}
        
        # Merge scan results
        for feature, files in feature_map.items():
            if feature not in registry["features"]:
                registry["features"][feature] = {"status": "active", "version": "0.1.0", "files": [], "dependencies": []}
            
            # Simple merge: add new files to the list
            current_files = set([f["path"] if isinstance(f, dict) else f for f in registry["features"][feature].get("files", [])])
            for file_path in files:
                current_files.add(file_path)
            
            # Reconstruct detailed file objects (simplified for now)
            registry["features"][feature]["files"] = sorted(list(current_files))
            registry["features"][feature]["last_synced"] = "manual_scan_timestamp"

        with open(REGISTRY_PATH, 'w') as f:
            json.dump(registry, f, indent=2)
        print("Registry Updated.")
    elif args.audit:
        changed_files = get_changed_files()
        
        # Flatten feature_map to get set of all currently tagged files
        all_tagged_files = set()
        for file_list in feature_map.values():
            all_tagged_files.update(file_list) # file_list contains string paths
            
        untracked = []
        for cf in changed_files:
            # Normalize path separators
            cf_str = str(cf).replace(os.sep, '/')
            # Check if this changed file appears in ANY feature list
            # We use loose matching because file_list might be relative differently? 
            # Actually scan_file uses relative paths so exact match should work if normalized.
            
            # Simple check: Is this file in the set of all feature-tagged files?
            # Windows path normalization is key here. scan_file uses Path(file_path) which might use backslashes.
            # feature_map stores str(path).
            
            # Use Path object comparison for robustness
            is_tagged = False
            for tagged_path in all_tagged_files:
                if Path(tagged_path).resolve() == cf.resolve():
                    is_tagged = True
                    break
            
            if not is_tagged:
                untracked.append(str(cf))
                
        if untracked:
            print(json.dumps({
                "status": "warning",
                "message": "Found modified/new files without Feature Tags.",
                "untracked_files": untracked
            }, indent=2))
            exit(1) # Signal failure to the agent
        else:
            print(json.dumps({"status": "success", "message": "All modified files are properly tagged."}, indent=2))
    else:
        print(json.dumps(feature_map, indent=2))

if __name__ == "__main__":
    main()

File: .portkit/scripts/typescript/update-registry.ts
// Scan Registry - Maintenance Tool
// Purpose: Scans codebase for @feature annotations and updates feature-registry.json.

import * as fs from 'fs';
import * as path from 'path';
import { execSync } from 'child_process';

const REGISTRY_PATH = path.join('.portkit', 'addon-features-registry', 'feature-registry.json');

interface FeatureMap {
    [key: string]: string[];
}

interface Registry {
    metadata: { version: string; last_updated: string };
    features: {
        [key: string]: {
            status: string;
            version: string;
            files: string[];
            dependencies: string[];
            last_synced: string;
        };
    };
}

function scanFile(filePath: string, featureMap: FeatureMap) {
    try {
        const content = fs.readFileSync(filePath, 'utf-8');
        const matches = [
            ...content.matchAll(/\/\/\s*feature-start:\s*([\w-]+)/g),
            ...content.matchAll(/#\s*feature-start:\s*([\w-]+)/g)
        ];

        for (const match of matches) {
            const feature = match[1];
            if (!featureMap[feature]) {
                featureMap[feature] = [];
            }
            // Normalize to POSIX
            const normalizedPath = filePath.split(path.sep).join('/');
            if (!featureMap[feature].includes(normalizedPath)) {
                featureMap[feature].push(normalizedPath);
            }
        }
    } catch (e) {
        // Ignore errors
    }
}

function getChangedFiles(): string[] {
    const files = new Set<string>();
    try {
        // 1. Modified tracked files
        const resMod = execSync('git diff --name-only HEAD', { encoding: 'utf-8' });
        resMod.split('\n').filter(Boolean).forEach(f => files.add(f.trim()));

        // 2. Untracked (new) files
        const resNew = execSync('git ls-files --others --exclude-standard', { encoding: 'utf-8' });
        resNew.split('\n').filter(Boolean).forEach(f => files.add(f.trim()));
    } catch (e) {
        console.error(`Warning: Could not get git status: ${e}`);
    }

    return Array.from(files).filter(f => /\.(ts|tsx|js|py|md)$/.test(f));
}

function main() {
    const args = process.argv.slice(2);
    const update = args.includes('--update');
    const audit = args.includes('--audit');

    // Parse --folder argument
    let scanFolder = '.';
    const folderIndex = args.indexOf('--folder');
    if (folderIndex !== -1 && args[folderIndex + 1]) {
        scanFolder = args[folderIndex + 1];
    }

    // Normalize path to ensure git command works well
    scanFolder = path.resolve(scanFolder);

    const featureMap: FeatureMap = {};

    try {
        // OPTIMIZATION: Use git grep to find ONLY files that contain our tags.
        // We limit search to the specified folder if provided.
        // ':/' makes path relative to repo root if using cwd, but let's be careful.
        // If we want to search a subdirectory, we pass it to git grep.

        // Relative path suitable for git grep
        const relPath = path.relative(process.cwd(), scanFolder) || '.';
        const searchPath = relPath === '.' ? '' : `${relPath}/`; // Trailing slash

        const cmd = `git grep -I -l "feature-start:" -- "${searchPath}"`;

        const result = execSync(cmd, { encoding: 'utf-8', stdio: ['ignore', 'pipe', 'ignore'] }); // pipe stdout, ignore stderr (exit 1 if not found is annoying otherwise)

        const files = result.split('\n').filter(Boolean);

        for (const file of files) {
            const normalizedFile = file.trim();
            // if scanning subdirectory, git grep outputs relative path from CWD usually if run from CWD
            scanFile(normalizedFile, featureMap);
        }

    } catch (e: any) {
        // git grep returns 1 if no matches found
        if (e.status !== 1) {
            console.error(`Error running git grep: ${e}. Falling back to slow walk (not implemented in TS version yet).`);
        }
    }

    if (update) {
        let registry: Registry = { metadata: { version: '1.0', last_updated: new Date().toISOString() }, features: {} };
        if (fs.existsSync(REGISTRY_PATH)) {
            registry = JSON.parse(fs.readFileSync(REGISTRY_PATH, 'utf-8'));
        }

        for (const [feature, files] of Object.entries(featureMap)) {
            if (!registry.features[feature]) {
                registry.features[feature] = { status: 'active', version: '0.1.0', files: [], dependencies: [], last_synced: '' };
            }

            const currentFiles = new Set(registry.features[feature].files.concat(files).map(f => f.split(path.sep).join('/')));
            registry.features[feature].files = Array.from(currentFiles).sort();
            registry.features[feature].last_synced = "manual_scan_timestamp";
        }

        fs.writeFileSync(REGISTRY_PATH, JSON.stringify(registry, null, 2));
        console.log("Registry Updated.");

    } else if (audit) {
        const changedFiles = getChangedFiles();
        const allTaggedFiles = new Set<string>();

        // We need to reload registry or scan ALL if audit is global. 
        // Audit implies checking if CHANGED files are covered. 
        // NOTE: If we only scanned a folder, we might miss tags elsewhere, but usually audit is run on repo root.
        // For robustness, audit should probably NOT limit by folder unless user intends to ONLY audit that folder.
        // But let's assume if audit is run, we used full scan or the changed files are what matters.

        // Actually, if we use featureMap from a PARTIAL scan, we might flag valid files as untagged if they are outside the scan folder.
        // However, the Python script does a FULL scan for audit.
        // Let's assume for this implementation that audit requires a full scan or relies on the registry + scan?
        // The python script does a fresh scan_file loop.

        // For safety, let's warn if folder is used with audit
        if (folderIndex !== -1) {
            console.warn("Warning: Using --folder with --audit may report false positives if untagged files are outside the folder.");
        }

        Object.values(featureMap).forEach(files => files.forEach(f => allTaggedFiles.add(f)));

        const untracked: string[] = [];
        for (const cf of changedFiles) {
            const normalizedCf = cf.split(path.sep).join('/');

            // Check if tagged (naive check: is exact path in our map?)
            // NOTE: Python used resolve(), here we use string match assuming git ls-files and git grep return canonical relative paths
            if (!allTaggedFiles.has(normalizedCf)) {
                // Double check with resolve to be sure
                const absCf = path.resolve(cf);
                let found = false;
                for (const tagged of allTaggedFiles) {
                    if (path.resolve(tagged) === absCf) {
                        found = true;
                        break;
                    }
                }
                if (!found) untracked.push(normalizedCf);
            }
        }

        if (untracked.length > 0) {
            console.log(JSON.stringify({
                status: "warning",
                message: "Found modified/new files without Feature Tags.",
                untracked_files: untracked
            }, null, 2));
            process.exit(1);
        } else {
            console.log(JSON.stringify({ status: "success", message: "All modified files are properly tagged." }, null, 2));
        }

    } else {
        console.log(JSON.stringify(featureMap, null, 2));
    }
}

main();
